sw/device/lib/testing/usb_testutils_streams.c

```diff
@@ -0,0 +1,488 @@
+
+      rxtx_lfsr = LFSR_ADVANCE(rxtx_lfsr);
+      rx_lfsr = LFSR_ADVANCE(rx_lfsr);
+      sp++;
+    }
+
+    // Update the LFSRs for the next packet
+    s->rxtx_lfsr = rxtx_lfsr;
+    s->rx_lfsr = rx_lfsr;
+  } else {
+    // In the event that we've received a zero-length data packet, we still
+    // must return the buffer to the pool
+    CHECK_DIF_OK(
+        dif_usbdev_buffer_return(usbdev->dev, usbdev->buffer_pool, &buf));
+  }
+}
+
+// Callback for successful buffer transmission
+static void strm_tx_done(void *stream_v, usb_testutils_xfr_result_t result) {
+  usbdev_stream_t *s = (usbdev_stream_t *)stream_v;
+  usb_testutils_streams_ctx_t *ctx = s->ctx;
+  usb_testutils_ctx_t *usbdev = ctx->usbdev;
+
+  // If we do not have at least one queued buffer then something has gone wrong
+  // and this callback is inappropriate
+  uint8_t tx_ep = s->tx_ep;
+  uint8_t nqueued = ctx->tx_bufs_queued[tx_ep];
+
+  if (s->verbose) {
+    LOG_INFO("strm_tx_done called. %u (%u total) buffers(s) are queued",
+             nqueued, ctx->tx_queued_total);
+  }
+
+  CHECK(nqueued > 0);
+
+  // Note: since buffer transmission and completion signalling both occur within
+  // the foreground code (polling, not interrupt-driven) there is no issue of
+  // potential races here
+
+  if (nqueued > 0) {
+    // Shuffle the buffer descriptions, without using memmove
+    for (unsigned idx = 1u; idx < nqueued; idx++) {
+      ctx->tx_bufs[tx_ep][idx - 1u] = ctx->tx_bufs[tx_ep][idx];
+    }
+
+    // Is there another buffer ready to be transmitted?
+    ctx->tx_queued_total--;
+    ctx->tx_bufs_queued[tx_ep] = --nqueued;
+
+    if (nqueued) {
+      CHECK_DIF_OK(
+          dif_usbdev_send(usbdev->dev, tx_ep, &ctx->tx_bufs[tx_ep][0u]));
+    }
+  }
+}
+
+// Callback for buffer reception
+static void strm_rx(void *stream_v, dif_usbdev_rx_packet_info_t packet_info,
+                    dif_usbdev_buffer_t buf) {
+  usbdev_stream_t *s = (usbdev_stream_t *)stream_v;
+  usb_testutils_streams_ctx_t *ctx = s->ctx;
+  usb_testutils_ctx_t *usbdev = ctx->usbdev;
+
+  CHECK(packet_info.endpoint == s->rx_ep);
+
+  // We do not expect to receive SETUP packets to this endpoint
+  CHECK(!packet_info.is_setup);
+
+  if (s->verbose) {
+    LOG_INFO("Stream %u: Received buffer of %u bytes(s)", s->id,
+             packet_info.length);
+  }
+
+  if (s->sending && s->generating) {
+    buffer_check(ctx, s, packet_info, buf);
+  } else {
+    // Note: this is useful for profiling the OUT performance on CW310
+    usb_testutils_ctx_t *usbdev = ctx->usbdev;
+
+    if (read_method != kReadMethodNone && packet_info.length > 0) {
+      alignas(uint32_t) uint8_t data[USBDEV_MAX_PACKET_SIZE];
+      size_t len = packet_info.length;
+      size_t bytes_read;
+
+      switch (read_method) {
+#if USBUTILS_MEM_FASTER
+        // Faster read performance using custom routine
+        case kReadMethodFaster: {
+          const uint32_t USBDEV_BUFFER_REG_OFFSET = 0x800U;
```

Comment:
Consider using an enum, a `#define`, or keep as a variable but follow naming conversion to define this constant as per the [style guide](https://opentitan.org/book/doc/contributing/style_guides/c_cpp_coding_style.html#constants-and-preprocessor-macros).

Reply:
It has that name because the code and naming convention matches that in the DIF implementation of this routine. This has now been superseded by an effort to implement the faster routines in the DIF itself, so I've removed the code from here.